internal/k8s/controllers/config_controller.go

```diff
@@ -40,23 +43,33 @@ const bgpExtrasConfigName = "bgpextras"
 
 type ConfigReconciler struct {
 	client.Client
-	Logger         log.Logger
-	Scheme         *runtime.Scheme
-	Namespace      string
-	Handler        func(log.Logger, *config.Config) SyncState
-	ValidateConfig config.Validate
-	ForceReload    func()
-	BGPType        string
-	currentConfig  *config.Config
+	Logger          log.Logger
+	Scheme          *runtime.Scheme
+	Namespace       string
+	Handler         func(log.Logger, *config.Config) SyncState
+	ValidateConfig  config.Validate
+	ForceReload     func()
+	BGPType         string
+	ConfigStateName string
+	currentConfig   *config.Config
 }
 
 func (r *ConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	return requestHandler(r, ctx, req)
 }
 
-var requestHandler = func(r *ConfigReconciler, ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+var requestHandler = func(r *ConfigReconciler, ctx context.Context, req ctrl.Request) (result ctrl.Result, retError error) {
 	level.Info(r.Logger).Log("controller", "ConfigReconciler", "start reconcile", req.String())
 	defer level.Info(r.Logger).Log("controller", "ConfigReconciler", "end reconcile", req.String())
+
+	var conditionErr error
+	defer func() {
+		if err := r.reportCondition(ctx, conditionErr); err != nil {
+			level.Error(r.Logger).Log("controller", "ConfigReconciler", "message", "failed to report condition", "error", err)
+			retError = errors.Join(retError, err)
```

Comment:
please let's add a comment here saying this is returned. Or let's call it handlerError. I am afraid this is hiding a bit the behavior.
Also, shouldn't we have naked returns? I think this is shadowed by the explicit returns we have

Reply:
yes, I was also unsure about. I was investigating that, according to https://go.dev/ref/spec#Defer_statements
```
/ f returns 42
func f() (result int) {
	defer func() {
		// result is accessed after it was set to 6 by the return statement
		result *= 7
	}()
	return 6
}
```

Also in another part
```
A "return" statement that specifies results sets the result parameters before any deferred functions are executed.
```
AFAICT no shadowing  should take place, but I can try to rename retError to handlerError (or reconcilerError or resultError?)